sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -329,6 +334,40 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
     }
   }
 
+  private def unwrapTimeStampToDate(
+      exp: BinaryComparison,
+      fromExp: Expression,
+      ts: Literal,
+      tz: Option[String],
+      evalMode: EvalMode.Value): Expression = {
+    val floorDate = Cast(ts, fromExp.dataType, tz, evalMode)
+    val equalToFloorDateExpr =
+      EqualTo(ts, Cast(floorDate, ts.dataType, tz, evalMode))
+    val equalToFloorDate = equalToFloorDateExpr.eval(EmptyRow).asInstanceOf[Boolean]
+    val ceilDate =
+      if (equalToFloorDate) {
+        floorDate
+      } else {
+        DateAdd(floorDate, Literal(1, IntegerType))
+      }
+
+    exp match {
+      case _: GreaterThan =>
+        GreaterThan(fromExp, floorDate)
+      case _: GreaterThanOrEqual =>
+        GreaterThanOrEqual(fromExp, ceilDate)
+      case _: EqualTo =>
+        And(EqualTo(fromExp, floorDate), EqualTo(fromExp, ceilDate))
```

Comment:
```suggestion
        if (isStartOfDay) EqualTo(fromExp, floorDate) else FalseLiteral
```

Reply:
Thanks @wangyum for your review.

If `isStartOfDay` is false, the optimizer will change this expression to `FalseLiteral` which results in incorrect result if `fromExp` is null.
So I use two EqualTo expressions to support nullable input expressions. During the execution phase, if `isStartOfDay` is true, the two EqualTo expressions can be reuse and no additional calculations are added.

Reply:
How about?
```scala
        if (isStartOfDay) {
          EqualTo(fromExp, floorDate)
        } else if (!fromExp.nullable) {
          FalseLiteral
        } else {
          And(EqualTo(fromExp, floorDate), EqualTo(fromExp, ceilDate))
        }
```

Reply:
Updated, thanks